Tickets/dm 22966

mxk62 opened the pull request:
Initial implementation of the handoff part of the buffer manager.

mxk62 committed:
Add a custom formatter to the logger

mxk62 committed:
Refactor handling logging options

mxk62 committed:
Add core classes and schema to package namespace

mxk62 committed:
Add validation logic

mxk62 committed:
Rewrite code to implement new logic

Previous version was determining what has changed in a source directory
by comparing its two snapshots made at different time points.  Creating
a snapshot relied heavily on stat() function which might lead to
"beating down" the file system during normal operations due to a
combination of the number of files in the monitored directory and/or
sampling frequency.

However, in the current design of the system:

1. Writing files to the buffer is almost instantaneous (The Archiver
   does not actually copy the files to the buffer, but creates hard
   links instead.
2. Files in the buffer are immutable (The link is created after any
   writing to the actual file is completed).
3. After successful transfer, files are moved from the buffer to the
   holding area (a different directory).

As a result, the program could be modified to use a much simpler logic
as the buffer either contain new files or files which were not
transferred successfully in the past and need to be transferred anyway.
That eliminated the need to create and maintained directory snapshots.

mxk62 committed:
Add some unit tests

Added unit test for Scanner and Porter classes.

mxk62 committed:
Fix parsing error

Specifying a user on the remote location had no effect.  User account
running the script locally was always used instead.  Fixed it.

mxk62 committed:
Rewrite Scanner and Porter as commands

Two core classes: Scanner and Porter were subclasses of the Thread
class.  However, they were not run as daemons. As a result they needed
to be created and initialized in each iteration of the main loop.  I
rewrote them as commands. That way they can be initialized just once and
then run repeatedly, possibly in threads.

mxk62 committed:
Use f-strings instead of format()

Replaced all messages using format() with equivalent f-strings.

mxk62 committed:
Improve logging

I modified the format of the log messages to include the name of the
module and added some debug messages to the execute() function.

mxk62 committed:
Remove unused code

While creating directories at the remote location, the Porter class was
designed to suppress execution errors of 'mkdir' command related to
attempts of creating already existing directories.  However, since the
'mkdir' is called with '-p' option no such errors will be raised anyway
rendering the error check completely useless.  Also, even if such errors
occurred, they would not be suppressed as the implemented logic was
incorrect.  As a result, I removed this error check entirely.

mxk62 committed:
Make sure file transfer is "atomic"

In its first iteration, ctrl_oods will be used to emulate ingest
service.  However, it was relies on the fact that that placing files in
the OODS buffer is an atomic operation.  To emulate such a behavior I
modified the code to initially transfer file to a temporary staging area
from which they are moved (once transfer is completed) the buffer
managed by ctrl_oods.  Also added some cleanup commands to keep buffer
and staging area free from "leftover" directories.

mxk62 committed:
Group commands based on side they operate on

Grouped together commands based on the sides they are mostly concern of
(handoff vs endpoint) and renamed modules accordingly.

mxk62 committed:
Make the main script executable

mxk62 committed:
Run cleanup in the background

Files which were successfully transferred are moved from the buffer to
the holding area.  In previous implementation they were moved after all
files scheduled for transfer were processed.  This could lead to a
significant inconsistency with regard to files in the buffer and files
which were already transferred in case of the program crashing during
processing of large amount of files.  To mitigate this risk I modified
the program to run the cleanup in a background thread and move the
successfully transferred files to the holding area practically
immediately.

mxk62 committed:
Run cleanup only if transfers occured

An attempt to remove empty directories in the staging area was made in
every iteration regardless to if any transfer (and having any chance for
these empty directories to exist) occured.  I modified the program to
make those attempts only after any files were actually transferred.

mxk62 committed:
Change some naming

Make minor changes with regard to the main script name and paths in the
configuration file.

mxk62 committed:
Improve comments in configuration file

Modified comments in the configuration file to use terminology used in
the code base.

mxk62 committed:
Change buffer cleanup policy

Directories in the buffer of the handoff site were removed as soon as
they became empty.  This policy could lead to a race condition between
the archiver creating a directory to write a file and the buffer manager
removing it.  I modified the cleanup policy to remove empty
directories in the buffer only if no file was written to them for a
a given period of time.

mxk62 committed:
Improve docstrings

Made some minor changes to classes' docstring to make it more clear and
accurate.

mxk62 committed:
Log critical errors in Mover class

Critical errors in the Mover class were only displayed on the
stdout/stderr, but not logged.  Fixed it.

mxk62 committed:
Refactor setting options for logger

There was a little bit of repeated code in the section of the code
setting options for the logger.  Made it simpler too avoid these
repetitions.

mxk62 committed:
Do not spawn transfer thread if queue is empty

Threads responsible for transferring files were spawned regardless of
the fact if any files were enqueued for transfer.  Made changes to
prevent the code from doing this.

mxk62 committed:
Change severity level for few log messages

mxk62 committed:
Add log message format to configuration file

mxk62 committed:
Formalize interface for commands

The commands all the implemented commands share common interface it was
not formally enforced.  To address this limitation, I added an abstract
base class with required interface and made existing commands concrete
classes of it.

mxk62 committed:
Remove Python bytecode

Removed Python bytecode which was accidently added to the repository
some time ago.

mxk62 committed:
Bring unit tests up to date with the code base

Existing unit tests were out of sync with the code base.  I updated
existing ones to work with modified commands.  Also added extra unit
tests to cover newly added commands.

MichelleGower commented on bin/transd.py:
```diff
@@ -0,0 +1,171 @@
+    options : dict, optional
+       Logger settings. If None (default), default settings will be used.
+    """
+    settings = {
```
May want to put comment here about using a dict for ease of updating default values.  Since the dict isn't used as a dict anywhere, I was wondering why options.get(X, default=Y) or separate function parameters with defaults weren't used.

mxk62 replied:
I'm slightly confused here. What do you mean by:

> Since the dict isn't used as a dict anywhere

If the dictionary `options` is provided, its key/value pairs will be used to change logger settings from their default values to ones specified by user (lines 46-47).

Regarding using `options.get()` approach, it implies that `options` is always a dictionary. It is not the case in the current implementation. If there are no logging related entries in application configuration file, `options` will be set to **None**.

Finally, I didn't use optional parameters in `set_logger()` as I'm not really sure at this stage how many aspect of logging we want to make configurable. If we end up with just a few (say 3), I have no strong objections of making them optional arguments of the `set_logger()`. However, if there will be more I would prefer to stick to having it just one argument encapsulating its entire configuration.

MichelleGower replied:
By using it as a dict, I was referring to the fact that other than using it for the update call, its only use is to access individual values (i.e., the dict isn't passed to any other function).  The options=None case is easy to remedy by a check early in the function and setting it to an empty dict if None.   I didn't specify that it had to be a code change but instead could just add a comment.

MichelleGower commented on bin/transd.py:
```diff
@@ -0,0 +1,171 @@
+        time.sleep(pause)
+
+
+if __name__ == "__main__":
```
When LSST-ifying the code in a later ticket, I think that driver code is now put in a "main" function for easier unit testing. (See daf_butler driver updates.)

MichelleGower commented on bin/transd.py:
```diff
@@ -0,0 +1,171 @@
+
+    # Read provided configuration or use the default one.
+    if args.config is None:
+        root = os.getcwd()
```
Is the "default one" the etc/transd.yml file from this package?  If so, wouldn't it be better to have root be relative to the path for this python file as opposed to CWD?

MichelleGower commented on bin/transd.py:
```diff
@@ -0,0 +1,171 @@
+    timeout = settings["timeout"]
+    pool_size = settings["transfer_pool"]
+    exp_time = settings["expiration_time"]
+
```
It would be good to have a short comment about what each one of these objects do.

MichelleGower commented on bin/transd.py:
```diff
@@ -0,0 +1,171 @@
+        start = time.time()
+        scanner.run()
+        end = time.time()
+        eta = end - start
```
Please pick another name other than eta since it is being used to describe already completed events (as opposed to estimating future time completions).

MichelleGower commented on dbb_buffer_mngr/endpoint.py:
```diff
@@ -0,0 +1,184 @@
+class Wiper(Command):
+    """Class representing a cleanup command.
+
+    When invoked the command will remove all empty directories at a given
```
Not clear whether this class was intended to be generic (comments) or specific to staging area (code).  Please fix to match which ever one was intended.

MichelleGower commented on dbb_buffer_mngr/endpoint.py:
```diff
@@ -0,0 +1,184 @@
+    """
+    logger.debug(f"Executing {cmd}.")
+
+    args = cmd.split()
```
When using split for shell commands, I normally see shlex.split().

MichelleGower commented on dbb_buffer_mngr/handoff.py:
```diff
@@ -0,0 +1,135 @@
+            path = config["buffer"]
+        except KeyError:
+            raise ValueError("Buffer not specified.")
+        if not os.path.exists(path) or not os.path.isdir(path):
```
Since in the end returning same ValueError, why explicitly call both os.path.exists as well as os.path.isdir?  According to python docs, os.path.isdir checks existing as well.

mxk62 replied:
I must have misinterpreted the docs. I will remove the redundant check.

MichelleGower commented on dbb_buffer_mngr/handoff.py:
```diff
@@ -0,0 +1,135 @@
+        for topdir, subdir, filenames in os.walk(self.root):
+            paths.extend([os.path.join(topdir, f) for f in filenames])
+        for path in paths:
+            tail = path.replace(self.root, "").lstrip(os.sep)
```
See os.path.relpath

mxk62 replied:
Aaa, that's what I was looking for! Must have completely missed it for some reason.

MichelleGower commented on dbb_buffer_mngr/handoff.py:
```diff
@@ -0,0 +1,135 @@
+        paths = []
+        for topdir, subdir, filenames in os.walk(self.root):
+            paths.extend([os.path.join(topdir, f) for f in filenames])
+        for path in paths:
```
Can't the work be done in single loop?

MichelleGower commented on dbb_buffer_mngr/handoff.py:
```diff
@@ -0,0 +1,135 @@
+            duration = now - mod_time
+            if duration > self.exp_time:
+                os.rmdir(path)
+                logger.debug(f"Directory '{path}' removed successfully.")
```
If not removed successfully, the program throws an exception that isn't caught?

mxk62 replied:
According to the official [documentation](https://docs.python.org/3.8/library/os.html#os.rmdir) `os.rmdir()` throws an exception only if the directory does not exist or is not empty. The loop iterates over empty directories which were found earlier so wrapping it in **try-except** block seemed redundant.

MichelleGower replied:
As a program that you want to always keep running, this seems like something to catch and keep running (i.e., it isn't immediately bad).  While it should never happen procedurally, I've seen where folks debugging problems have made temporary files in their cwd while being in a wrong directory or done other things by hand that can cause problems with automated processes.

mxk62 replied:
You got a point, such scenarios didn't cross my mind. I'll add **try-except** block around it.

MichelleGower commented on tests/test_wiper.py:
```diff
@@ -0,0 +1,53 @@
+    def testNonEmpty(self):
+        subdir = tempfile.mkdtemp(dir=self.dir)
+        files = dict([tempfile.mkstemp(dir=os.path.join(self.dir, subdir))])
+        for fd in files:
```
Doesn't the call to mkstemp only create 1 file?

mxk62 replied:
Yes, it does. I just reused the snippet from other test which deals with multiple files. I will change it to reflect that it's just a one file.

MichelleGower commented on tests/test_wiper.py:
```diff
@@ -0,0 +1,53 @@
+
+        dirs = []
+        for top, subs, files in os.walk(self.dir):
+            for d in subs:
```
Seems strange that this isn't also checking for files, but I guess it works because the test as written only makes files in a subdir.

mxk62 replied:
The intention behind this test is to check if the command does not remove directories containing files. I don't really care about the file I created. I'm interested if the directory is still there after the command run.

MichelleGower replied:
I got myself confused about what Wiper does.  I thought this test was checking to make sure it was removing everything given a non-empty directory when Wiper is only supposed to get rid of empty directories.  Checking the assertEqual function at the end I see that the test does only succeed if it Wiper didn't delete the non-empty directory.   It would be helpful to have a short comment describing what each test is doing at the top of each test.

MichelleGower reviewed (changes requested):
Reviewed under the agreement that updating the code to project standards will be done under a later ticket branch.

mxk62 committed:
Refactor Scanner command

Replaced "manual" path manipulations with os.path.relpath(), removed
a redundant check, and make cleaning up the buffer more robust.

mxk62 committed:
Make finding default config file more robust

Defult settings for the location of the configuration file implicitly
assumed that the main application is being run from project's directory.
I modified the code to use path relative to the location of the script
instead.

mxk62 committed:
Improve docstrings and comments

mxk62 committed:
Use shlex to deal with shell commands

The string representing a shell command was split into options and
arguments with plain string split() method.  Replaced it with more
robust shlex.split().
